src/DonationForms/resources/app/DonationFormApp.tsx

```diff
@@ -12,70 +12,118 @@ import DonationFormErrorBoundary from '@givewp/forms/app/errors/boundaries/Donat
 import MultiStepForm from '@givewp/forms/app/form/MultiStepForm';
 import getDonationFormNodeSettings from '@givewp/forms/app/utilities/getDonationFormNodeSettings';
 import {DonationFormSettingsProvider} from '@givewp/forms/app/store/form-settings';
+import usePubSub from '@givewp/forms/app/utilities/usePubSub';
+import {useState, useEffect} from 'react';
+import {FormSettings} from '@givewp/form-builder/types';
 
 const formTemplates = window.givewp.form.templates;
 const GoalAchievedTemplate = withTemplateWrapper(formTemplates.layouts.goalAchieved);
 
 /**
  * Get data from the server
  */
-const {form} = getWindowData();
+const {form: initialFormState} = getWindowData();
+const donationFormNodeSettings = getDonationFormNodeSettings(initialFormState);
 
-prepareFormData(form);
+prepareFormData(initialFormState);
 
 mountWindowData();
 
 /**
  * Prepare default values for form
  */
-const defaultValues = getDefaultValuesFromSections(form.nodes);
+const defaultValues = getDefaultValuesFromSections(initialFormState.nodes);
 
-const schema = getJoiRulesForForm(form);
+const schema = getJoiRulesForForm(initialFormState);
 
 const initialState = {
     defaultValues,
     gateways: window.givewp.gateways.getAll(),
     validationSchema: schema,
 };
 
-const formSettings = {...form.settings, ...getDonationFormNodeSettings(form)};
-
 /**
  * @since 3.0.0
  */
-function App() {
+function App({preview}) {
+
+    const {subscribe} = usePubSub();
+    const [form, setFormState] = useState(initialFormState);
+
+    useEffect(() => {
+        if (preview) {
+            subscribe('preview:settings', (data: FormSettings) => {
+                setFormState(prevState => {
+                    return {
+                        ...prevState,
+                        settings: {
+                            ...prevState.settings,
+                            ...data
+                        }
+                    }
+                })
+            })
+
+            subscribe('preview:goal', (data: { [s: string]: string }) => {
+                setFormState(prevState => {
+                    return {
+                        ...prevState,
+                        goal: {
+                            ...prevState.goal,
+                            ...data
+                        }
+                    }
+                })
+            })
+
+            subscribe('preview:colors', (data: { [s: string]: string }) => {
+                const [key, value] = Object.entries<string>(data).flat();
```

Comment:
Let's take this in a slightly different direction. Reason being, right now regardless of preview mode, we're duplicating the form object, loading the subscribe function, and adding a useEffect.  As you can probably guess, I want to be extra careful about preview mode leaking into live mode of the form.

 Instead of adding `preview` as a prop let's actually create an `<AppPreview />` component who's sole purpose is to load the subscribe function and use effect.  It then simply returns `<App form={previewForm} />`.  The form prop would be an optional new prop on the App component so it can come from the server or be overwritten by the prop.
 
 So down below we would do `render(preview ? <AppPreview /> : <App />, root);`

Reply:
Make sense

Reply:
@alaca @jonwaldstein I actually _do_ want a concept of preview within the app. The reason being I want to do more with this in subsequent PRs. For example, in preview mode I don't want it to attempt actual donations. Clicking "donate" should just display the confirmation receipt with the entered data, but not actually attempt to make a donation.

Reply:
@JasonTheAdams we can add that back. I didn't delete it. What do you think about having that information in form state? Then we don't have to pass it everywhere as a prop.

Reply:
Oh yeah, for sure. We want various parts of the form to be able to act a bit differently during preview. Preview mode won't change back and forth, so I'm not worried about it causing unnecessary re-renders.

Reply:
Resolved https://github.com/impress-org/givewp/pull/7059/commits/238ade6ca2ec1f7e5c082e0d7a8c2a31e61694e3

Reply:
I'm fine with a top level preview flag in the form context like @alaca suggested.  I just want to be careful about preview functionality being loaded inside the donation form when it's not in preview mode.